### PR TITLE
Martien.kissfix

### PR DESCRIPTION
**Pull request by martien-de-jong**
dup phi elimination crashed in a supertest on a block with just two phi nodes

### CODE REVIEW

File: llvm/lib/Target/AIE/AIEEliminateDuplicatePHI.cpp [hunk 1]
```diff
@@ -120,26 +120,26 @@ class AIEEliminateDuplicatePHI : public MachineFunctionPass {
...
     bool Changed = false;
-    auto LastPHI = MBB.getFirstNonPHI();
-    for (auto &MI : make_early_inc_range(MBB.phis())) {
-      if (!isPointerTypePHI(MRI, MI))
-        continue;
-
-      // Ensure that next MI is valid
-      auto NextMI = std::next(MachineBasicBlock::instr_iterator(MI));
-      if (NextMI == MBB.instr_end() || NextMI == LastPHI)
+    const auto End = MBB.end();
```
Comment (martien-de-jong) [inline llvm/lib/Target/AIE/AIEEliminateDuplicatePHI.cpp:123]: We are comparing every pointer PHI node in the block to any other pointer PHI node in that block.  We know that all PHI nodes are consecutive at the start of the block. The transformation only changes operands in other instructions, so iterators stay intact. By writing the loops using basic iterator idiom reveals the symmetry between the two loops. It also avoids using constructs like .phis() and .getFirstNonPhi() that both perform exactly the same iteration as the main loops.

File: llvm/lib/Target/AIE/AIEEliminateDuplicatePHI.cpp [hunk 2]
```diff
@@ -120,26 +120,26 @@ class AIEEliminateDuplicatePHI : public MachineFunctionPass {
...
+    for (auto PhiItA = MBB.begin(); PhiItA != End; ++PhiItA) {
+      auto &PhiA = *PhiItA;
+      if (PhiA.getOpcode() != TargetOpcode::G_PHI)
         break;
-
-      for (auto &PHI : make_early_inc_range(make_range(
-               NextMI, MachineBasicBlock::instr_iterator(*LastPHI)))) {
```
Comment (martien-de-jong) [inline llvm/lib/Target/AIE/AIEEliminateDuplicatePHI.cpp:134]: The error was here. LastPHI was at .end(), which means you cannot dereference it to create an instr_iterator. Note that in that case the validity check reads `if (NextMI == end || NextMI == end)`

File: llvm/lib/Target/AIE/AIEEliminateDuplicatePHI.cpp [hunk 3]
```diff
@@ -120,26 +120,26 @@ class AIEEliminateDuplicatePHI : public MachineFunctionPass {
   bool processMBB(MachineBasicBlock &MBB, MachineRegisterInfo &MRI,
                   MachineIRBuilder &MIB, GISelObserverWrapper &Observer) {
     bool Changed = false;
-    auto LastPHI = MBB.getFirstNonPHI();
```
Comment (martien-de-jong) [inline llvm/lib/Target/AIE/AIEEliminateDuplicatePHI.cpp:123]: I would normally expect an iterator with this name to point to the last phi, not the the first non-phi.

File: llvm/test/CodeGen/AIE/GlobalISel/eliminate-duplicate-phi.mir [hunk 4]
```diff
@@ -232,3 +232,51 @@ body:             |
...
+
+  bb.1:
+    G_BR %bb.3
+
+  bb.2:
+    G_BR %bb.3
+
+  bb.3:
+    %3:_(p0) = G_PHI %2:_(p0), %bb.1, %3:_(p0), %bb.2
+    %4:_(p0) = G_PHI %2:_(p0), %bb.1, %3:_(p0), %bb.2
```
Comment (martien-de-jong) [inline llvm/test/CodeGen/AIE/GlobalISel/eliminate-duplicate-phi.mir:284]: This is just triggering the crash. These are equivalent phis, and the test can be strengthened by using %4 in a downstream block.

File: llvm/lib/Target/AIE/AIEEliminateDuplicatePHI.cpp [hunk 5]
```diff
@@ -120,26 +120,26 @@ class AIEEliminateDuplicatePHI : public MachineFunctionPass {
...
+      if (PhiA.getOpcode() != TargetOpcode::G_PHI)
         break;
-
-      for (auto &PHI : make_early_inc_range(make_range(
-               NextMI, MachineBasicBlock::instr_iterator(*LastPHI)))) {
-        if (!isPointerTypePHI(MRI, PHI))
+      if (!isPointerTypePHI(MRI, PhiA))
+        continue;
+      auto PhiItB = PhiItA;
+      for (++PhiItB++; PhiItB != End; ++PhiItB) {
```
Comment (andcarminati) [inline llvm/lib/Target/AIE/AIEEliminateDuplicatePHI.cpp]: Looks interesting: `++PhiItB++`

**Comment by martien-de-jong** [comment]
I approved the competing PR

**Pull request closed without merging**